File: src/p2.js
import * as R from "rambda";

//given in question
const seqLimit = 4_000_000;

//generate the next fibonacci

const nextFibo = ([a, b]) => [b, a + b];

//check if even
const isEven = (n) => n % 2 === 0;

// generate fibonacci to the given limit
const fibsUpTo = (limit) => {
  const genFibs = ([a, b]) =>
    a > limit ? [] : [a, ...genFibs(nextFibo([a, b]))];

  return genFibs([1, 2]);
};

//check if a number is less than or equal to the limit
const isLessThanOrEqual = (limit) => (n) => n <= limit;

//handler to the methods aka main
const sumEvenFibos = R.pipe(
  fibsUpTo,
  R.takeWhile(isLessThanOrEqual(seqLimit)),
  R.filter(isEven),
  R.sum
);

//result
const result = sumEvenFibos(seqLimit);

console.log(result);

File: src/p2.ml
let seqLimit = 4_000_000

let isEven n = 
  if n mod 2 = 0 then true else false 

let isLessThanOrEqual limit n = 
  n <= limit 

let nextFibo (a,b) = (b,a+b)

let rec fibsUpTo limit (a,b) =
  if a > limit then []
  else a :: fibsUpTo limit (nextFibo (a,b))

let sumEvenFibos limit =
  fibsUpTo limit (1,2)
  |> List.filter isEven
  |> List.filter (isLessThanOrEqual limit)
  |> List.fold_left ( + ) 0

let result = 
  sumEvenFibos seqLimit

let () =
  Printf.printf "%d\n" result 

